File: exp-11/q1-dijkstra/solution.cpp
#include <iostream>
using namespace std;

const int MAX = 100;
const int INF = 1000000000;

int main()
{
    int n, m, s;
    cin >> n >> m;

    int adj[MAX][MAX];
    for (int i = 1; i <= n; i++)
        for (int j = 1; j <= n; j++)
            adj[i][j] = INF;

    for (int i = 0; i < m; i++)
    {
        int u, v, w;
        cin >> u >> v >> w;
        adj[u][v] = w;
    }
    cin >> s;

    int dist[MAX], visited[MAX];
    for (int i = 1; i <= n; i++)
    {
        dist[i] = INF;
        visited[i] = 0;
    }
    dist[s] = 0;

    for (int count = 0; count < n; count++)
    {
        int u = -1, min = INF;
        for (int i = 1; i <= n; i++)
        {
            if (!visited[i] && dist[i] < min)
            {
                min = dist[i];
                u = i;
            }
        }
        if (u == -1)
            break;
        visited[u] = 1;
        for (int v = 1; v <= n; v++)
        {
            if (!visited[v] && adj[u][v] < INF &&
                dist[u] + adj[u][v] < dist[v])
            {
                dist[v] = dist[u] + adj[u][v];
            }
        }
    }

    for (int i = 1; i <= n; i++)
        cout << "Distance from " << s << " to " << i << ": " << dist[i] << endl;
    return 0;
}

File: exp-11/q1-dijkstra/solution.js
const fs = require("fs");

const INF = 1000000000;
const MAX = 100;

const data = fs.readFileSync(0, "utf-8").trim().split(/\s+/).map(Number);
let idx = 0;
const n = data[idx++];
const m = data[idx++];

const adj = Array.from({ length: MAX }, () => new Array(MAX).fill(INF));
for (let i = 0; i < m; i++) {
  const u = data[idx++];
  const v = data[idx++];
  const w = data[idx++];
  adj[u][v] = w;
}
const s = data[idx++];

const dist = new Array(MAX).fill(INF);
const visited = new Array(MAX).fill(false);
dist[s] = 0;

for (let count = 0; count < n; count++) {
  let u = -1;
  let min = INF;
  for (let i = 1; i <= n; i++) {
    if (!visited[i] && dist[i] < min) {
      min = dist[i];
      u = i;
    }
  }
  if (u === -1) break;
  visited[u] = true;
  for (let v = 1; v <= n; v++) {
    if (!visited[v] && adj[u][v] < INF && dist[u] + adj[u][v] < dist[v]) {
      dist[v] = dist[u] + adj[u][v];
    }
  }
}

for (let i = 1; i <= n; i++) {
  console.log(`Distance from ${s} to ${i}: ${dist[i]}`);
}
